tests/integration/kubernetes/lib.sh

```diff
@@ -81,6 +81,61 @@ assert_logs_contain() {
 	# Note: with image-rs we get more than the default 1000 lines of logs
 	print_node_journal "$node" "$log_id" --since '"'$datetime'"' | grep "$message"
 }
+
+# Function to check logs every 10 seconds
+check_logs() {
+	local node="$1"
+	local tag="$2"
+	local node_start_time="$3"
+	local log_message="$4"
+	local duration="${5:-120}"
+
+	# Initialize the result variable to a non-zero value
+	local result=1
+	local elapsed_time=0
+	local sleep_interval=10
+
+	# Loop to run assert_logs_contain every 10 seconds
+	set +e
+	while [ $elapsed_time -lt $duration ]; do
+		assert_logs_contain "$node" "$tag" "$node_start_time" "$log_message"
+
+		# Check the return code of the last command
+		if [ $? -eq 0 ]; then
+			result=0  # Set result to 0 if assert_logs_contain succeeds
+			# check if a pod is not running
+			pod_name=$(kubectl get pods -o jsonpath='{.items..metadata.name}')
+			if ! kubectl get pods "$pod_name" | grep -q "Running"; then
+				echo "Logs found and the pod is not in a running state."
+				break
+			else
+				echo "Logs found, but the pod is not in a failed state."
+			fi
+		fi
+
+		sleep $sleep_interval
+		elapsed_time=$((elapsed_time + sleep_interval))
+	done
+	set -e
+	[ "$result" -ne 1 ]
+}
+
+# Create a pod then assert it fails to run
+# and check the logs for a given message.
+assert_logs_pod_fail() {
+	local pod_config="$1"
+	local node="$2"
+	local tag="$3"
+	local node_start_time="$4"
+	local log_message="$5"
+
+	assert_pod_fail "$pod_config" &
+	assert_pod_fail_pid=$!
+
+	check_logs "$node" "${tag}" "$node_start_time" "$log_message"
+	wait $assert_pod_fail_pid
```

Comment:
Hi @BbolroC ,

Suppose that check_logs() returned true after 30min then it will hit this line and wait  `assert_pod_fail`. The time for this test to finish is still 2min. am I missing something?

If I'm right, I was thinking in implement something like:
* do not run assert_pod_fail (not even on background)
* run check_logs
  * on check_logs true, check one-time pod is not Ready
  * on check_logs false, the test failed and it will return immediately

One warn for my idea: it's too fragile. Any bug on `check_logs` will likely to make the test always to pass (check_logs returns and the one-time check passes too because the pod didn't have time to be Ready, but it might get ready later).

Reply:
Yes and No:

Yes: the background job logic does follow your suggestion, `assert_pod_fail()` reaches the `wait` immediately after https://github.com/kata-containers/kata-containers/blob/8a893cd4ee274d3f457fc13c2aaabb6532211f01/tests/integration/kubernetes/lib.sh#L52 finishes. At that point, the test depends solely on `check_logs()`. If the message is captured and the pod isn’t in a running state, the test passes. Otherwise, it fails if the message does not appear within 120s (the default duration for `check_logs()`). But, of course, we can make `assert_logs_pod_fail()` synchronous as suggested unless `assert_pod_fail()` is used - which incurs an unnecessary 2-min wait due to `k8s_create_pod()`.

No: I don't think there is any chance a pod that produces an error message would end up in a ready state (even the negative test is not designed to have a pod ready) But you are saying the test should check that an error message is captured and ensure a pod doesn't get ready at least within 2 mins. Then yes, keeping the current logic would be better. 😉 

If you are still concerned about a background job, I can change the logic as you suggested. Just say the word, my lord. Thanks.